Python/tier2_redundancy_eliminator_bytecodes.c

```diff
@@ -189,6 +177,58 @@ dummy_func(void) {
         }
     }
 
+    op(_TO_BOOL, (value -- res)) {
+        (void)value;
+        OUT_OF_SPACE_IF_NULL(res = sym_new_known_type(ctx, &PyBool_Type));
+    }
+
+    op(_TO_BOOL_BOOL, (value -- value)) {
+        if (sym_matches_type(value, &PyBool_Type)) {
+            REPLACE_OP(this_instr, _NOP, 0, 0);
+        }
+        else {
+            sym_set_type(value, &PyBool_Type);
+        }
+    }
+
+    op(_TO_BOOL_INT, (value -- res)) {
+        sym_set_type(value, &PyLong_Type);
+        if (is_const(value)) {
+            PyObject *load = _PyLong_IsZero((PyLongObject *)get_const(value))
+                             ? Py_False : Py_True;
+            REPLACE_OP(this_instr, _POP_TOP_LOAD_CONST_INLINE_BORROW, 0, (uintptr_t)load);
+            OUT_OF_SPACE_IF_NULL(res = sym_new_const(ctx, load));
+        }
+        else {
+            OUT_OF_SPACE_IF_NULL(res = sym_new_known_type(ctx, &PyBool_Type));
+        }
+    }
+
+    op(_TO_BOOL_LIST, (value -- res)) {
+        sym_set_type(value, &PyList_Type);
+        OUT_OF_SPACE_IF_NULL(res = sym_new_known_type(ctx, &PyBool_Type));
+    }
+
+    op(_TO_BOOL_NONE, (value -- res)) {
+        if (get_const(value) == Py_None) {
+            REPLACE_OP(this_instr, _POP_TOP_LOAD_CONST_INLINE_BORROW, 0, (uintptr_t)Py_None);
+        }
```

Comment:
If feels like there should be an `else` block that raises some kind of error if it's a constant but it isn't None. Or maybe there should be a check at the top that it's either unknown or None, and nothing else?

Reply:
There is no error, but maybe we should optimize it differently.
Symbolic values aren't real values, so different assumptions apply.

Clearly we need to document this properly, or it is going to keep tripping people up. 
https://github.com/python/cpython/issues/115816